builder/builder.go

```diff
@@ -0,0 +1,457 @@
+	if head, err := b.bc.HeadsHeader(); err == nil {
+		expectedParentHash = head.Hash
+	}
+	if p.Block.ParentHash.Equal(expectedParentHash) {
+		return p, nil
+	}
+
+	return nil, ErrPendingParentHash
+}
+
+func (b *Builder) PendingBlock() *core.Block {
+	pending, err := b.Pending()
+	if err != nil {
+		return nil
+	}
+	return pending.Block
+}
+
+func (b *Builder) PendingState() (core.StateReader, func() error, error) {
+	txn, err := b.db.NewTransaction(false)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	pending, err := b.Pending()
+	if err != nil {
+		return nil, nil, utils.RunAndWrapOnError(txn.Discard, err)
+	}
+
+	return sync.NewPendingState(pending.StateUpdate.StateDiff, pending.NewClasses, b.headState), txn.Discard, nil
+}
+
+func (b *Builder) Run(ctx context.Context) error {
+	defer func() {
+		b.mempool.Close()
+	}()
+	// Clear pending state on shutdown
+	defer func() {
+		if pErr := b.ClearPending(); pErr != nil {
+			b.log.Errorw("clearing pending", "err", pErr)
+		}
+	}()
+
+	if err := b.InitPendingBlock(); err != nil {
+		return err
+	}
+
+	doneListen := make(chan struct{})
+	go func() {
+		if pErr := b.listenPool(ctx); pErr != nil {
+			if pErr != mempool.ErrTxnPoolEmpty {
+				b.log.Warnw("listening pool", "err", pErr)
+			}
+		}
+		close(doneListen)
+	}()
+
+	for {
+		select {
+		case <-ctx.Done():
+			<-doneListen
+			return nil
+		case <-time.After(b.blockTime):
+			b.finaliseMutex.Lock()
+			err := b.Finalise(b.Sign)
+			b.log.Infof("Finalised new block")
+			b.finaliseMutex.Unlock()
+			if err != nil {
+				return err
+			}
+		}
+	}
+}
+
+func (b *Builder) ClearPending() error {
+	b.pendingBlock.Store(&sync.Pending{})
+
+	if b.headState != nil {
+		if err := b.headCloser(); err != nil {
+			return err
+		}
+		b.headState = nil
+		b.headCloser = nil
+	}
+	return nil
+}
+
+func (b *Builder) InitPendingBlock() error {
+	header, err := b.bc.HeadsHeader()
+	if err != nil {
+		return err
+	}
+	pendingBlock := core.Block{
+		Header: &core.Header{
+			ParentHash:       header.Hash,
+			Number:           header.Number + 1,
+			SequencerAddress: &b.ownAddress,
+			// Todo: dynamically set fees (follow-up PR)
+			L1GasPriceETH:  felt.One.Clone(),
+			L1GasPriceSTRK: felt.One.Clone(),
+			L1DAMode:       core.Calldata,
+			L1DataGasPrice: &core.GasPrice{
+				PriceInWei: felt.One.Clone(),
+				PriceInFri: felt.One.Clone(),
+			},
+		},
+		Transactions: []core.Transaction{},
+		Receipts:     []*core.TransactionReceipt{},
+	}
+	newClasses := make(map[felt.Felt]core.Class)
+	emptyStateDiff := core.EmptyStateDiff()
+	su := core.StateUpdate{
+		StateDiff: &emptyStateDiff,
+	}
+	pending := sync.Pending{
+		Block:       &pendingBlock,
+		StateUpdate: &su,
+		NewClasses:  newClasses,
+	}
+	b.pendingBlock.Store(&pending)
+	b.headState, b.headCloser, err = b.bc.HeadState()
+	return err
+}
+
+// Finalise the pending block and initialise the next one
+func (b *Builder) Finalise(signFunc blockchain.BlockSignFunc) error {
```

Comment:
call it `sign` only to match it's previous name or call the others `signFunc`. The name is not the problem, is referring to it differently in different contexts (small nit)